File: WebClient SDK/src/default/ui/krProcess/krRecalcStagesUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext, CardModelFlags } from 'tessa/ui/cards';
import {
  getKrComponentsByCard,
  KrTypesCache,
  KrComponents,
  KrToken,
  CanFullRecalcRoute
} from 'tessa/workflow';
import { hasNotFlag, DotNetType, createTypedField } from 'tessa/platform';
import { GridViewModel } from 'tessa/ui/cards/controls';
import { UIButton } from 'tessa/ui';
import { IStorage } from 'tessa/platform/storage';

export class KrRecalcStagesUIExtension extends CardUIExtension {
  public initialized(context: ICardUIExtensionContext) {
    const cardModel = context.model;

    const usedComponents = getKrComponentsByCard(cardModel.card, KrTypesCache.instance);
    // Выходим если нет согласования
    if (hasNotFlag(usedComponents, KrComponents.Routes)) {
      return;
    }

    const approvalStagesTable = cardModel.controls.get('ApprovalStagesTable') as GridViewModel;
    let krToken: KrToken | null = null;
    if (
      approvalStagesTable &&
      hasNotFlag(cardModel.flags, CardModelFlags.Disabled) &&
      (krToken = KrToken.tryGet(context.card.info)) &&
      krToken.hasPermission(CanFullRecalcRoute)
    ) {
      const uiContext = context.uiContext;

      approvalStagesTable.leftButtons.push(
        UIButton.create({
          name: 'Recalc',
          caption: '$CardTypes_Buttons_RecalcApprovalStages',
          buttonAction: () => {
            const editor = uiContext.cardEditor;
            if (editor && !editor.operationInProgress) {
              const info: IStorage = {
                '.Recalc': createTypedField(true, DotNetType.Boolean)
              };
              editor.saveCard(uiContext, info);
            }
          }
        })
      );
    }
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrProcess/KrRecalcStagesUIExtension.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared.Workflow.KrPermissions;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;

namespace Tessa.Extensions.Default.Client.UI.KrProcess
{
    /// <summary>
    /// Расширение для модели представления карточки, добавляющее кнопку для пересчёта маршрута ("Пересчитать").
    /// </summary>
    public sealed class KrRecalcStagesUIExtension :
        CardUIExtension
    {
        #region Fields

        private readonly IKrTypesCache typesCache;

        #endregion

        #region Constructors

        public KrRecalcStagesUIExtension(IKrTypesCache typesCache) =>
            this.typesCache = typesCache ?? throw new ArgumentNullException(nameof(typesCache));

        #endregion

        #region Base Overrides

        /// <inheritdoc/>
        public override async Task Initialized(ICardUIExtensionContext context)
        {
            // выходим, если не включены процессы маршрутов
            var cardModel = context.Model;

            if (cardModel.Flags.HasAny(CardModelFlags.Disabled)
                || cardModel.Controls.TryGet(KrConstants.Ui.KrApprovalStagesControlAlias) is not GridViewModel approvalStagesTable
                || (await KrComponentsHelper.GetKrComponentsAsync(
                        cardModel.Card,
                        this.typesCache,
                        context.CancellationToken)).HasNot(KrComponents.Routes)
                || (KrToken.TryGet(context.Card.Info)?.HasPermission(KrPermissionFlagDescriptors.CanFullRecalcRoute)) != true)
            {
                return;
            }

            var uiContext = context.UIContext;

            approvalStagesTable.LeftButtons.Add(
                new UIButton("$CardTypes_Buttons_RecalcApprovalStages",
                    async _ =>
                    {
                        var editor = uiContext.CardEditor;
                        if (editor?.OperationInProgress != false)
                        {
                            return;
                        }

                        var info = new Dictionary<string, object>(StringComparer.Ordinal);
                        info.SetRecalcFlag();

                        await editor.SaveCardAsync(uiContext, info);
                    }));
        }

        #endregion
    }
}
